cnd/src/db/wrapper_types/mod.rs

```diff
@@ -112,3 +115,68 @@ impl From<ethereum::Address> for EthereumAddress {
         EthereumAddress(address)
     }
 }
+
+/// A wrapper type for Bitcoin networks.
+///
+/// This is then wrapped in the db::custom_sql_types::Text to be stored in DB
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum BitcoinNetwork {
+    Mainnet,
+    Testnet,
+    Regtest,
+}
+
+#[derive(Debug, Clone, Copy, thiserror::Error)]
+pub enum Error {
+    #[error("Unknown variant")]
+    Unknown,
+}
+
+impl FromStr for BitcoinNetwork {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "mainnet" => Ok(Self::Mainnet),
+            "testnet" => Ok(Self::Testnet),
+            "regtest" => Ok(Self::Regtest),
+            _ => Err(Error::Unknown),
+        }
+    }
+}
+
+impl fmt::Display for BitcoinNetwork {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = match self {
+            Self::Mainnet => "mainnet",
+            Self::Testnet => "testnet",
+            Self::Regtest => "regtest",
+        };
+        write!(f, "{}", s)
+    }
+}
+
+macro_rules! impl_from_for_bitcoinnetwork {
+    ($ledger:ident) => {
+        impl From<$ledger> for BitcoinNetwork {
+            fn from(_: $ledger) -> Self {
+                BitcoinNetwork::$ledger
+            }
+        }
+    };
+}
+
+use ledger::bitcoin::{Mainnet, Regtest, Testnet};
+impl_from_for_bitcoinnetwork!(Mainnet);
+impl_from_for_bitcoinnetwork!(Testnet);
+impl_from_for_bitcoinnetwork!(Regtest);
+
+impl From<BitcoinNetwork> for LedgerKind {
+    fn from(network: BitcoinNetwork) -> Self {
+        match network {
+            BitcoinNetwork::Mainnet => LedgerKind::BitcoinMainnet,
+            BitcoinNetwork::Testnet => LedgerKind::BitcoinTestnet,
+            BitcoinNetwork::Regtest => LedgerKind::BitcoinRegtest,
```

Comment:
These are going, IMO, to be hard to use because of the difference in format.  `Mainnet` is separate for one type and 'attached' for the other (`BitcoinMainnet`).  Is there a way to have these more uniform?

Reply:
And later we have `ledger::bitcoin::Testnet`

Perhaps we could use a nested enum:
```
enum LedgerKind {
    enum Bitcoin {
		Regtest,
		Mainnet,
		Testnet,
	},
	Ethereum,
}
``` 
Haven't actually done this before, I just looked it up.

Reply:
The nested enum may actually save some code dupe when doing matching. I did consider it but wasn't sure. I'll try and see how it turns out.

Reply:
> And later we have `ledger::bitcoin::Testnet`
> 
> Perhaps we could use a nested enum:
> 
> ```
> enum LedgerKind {
>     enum Bitcoin {
> 		Regtest,
> 		Mainnet,
> 		Testnet,
> 	},
> 	Ethereum,
> }
> ```
> 
> Haven't actually done this before, I just looked it up.

Where did you look it up? My compiler does not like the syntax.

Reply:
I checked, each enum needs to be defined separately:
```
pub enum BitcoinKind {
	Mainnet,
	Testnet,
	Regtest
}

pub enum LedgerKind {
	Bitcoin(BitcoinKind),
	Ethereum
}